lib/Dialect/QUIR/Transforms/MergeCircuits.cpp

```diff
@@ -259,9 +293,119 @@ MergeCircuitsPass::getCircuitOp(CallCircuitOp callCircuitOp,
+    auto arg = nextCircuitOp.getArgument(cnt);
+    int argumentIndex = 0;
+    if (find(insertedArguments, cnt) != insertedArguments.end()) {
+      auto dictArg = nextCircuitOp.getArgAttrDict(cnt);
+      newCircuitOp.insertArgument(baseArgNum + insertedCount, arg.getType(),
+                                  dictArg, arg.getLoc());
+      argumentIndex = baseArgNum + insertedCount;
+      insertedCount++;
+    } else {
+      argumentIndex = reusedArguments[cnt];
+    }
+    mapper.map(arg, newCircuitOp.getArgument(argumentIndex));
+  }
+}
+
+int MergeCircuitsPass::mapBarrierOperands(
+    Operation *barrierOp, CircuitOp newCircuitOp,
+    llvm::SmallVector<int> &insertedArguments,
+    std::unordered_map<int, int> &reusedArguments, IRMapping &mapper,
+    MLIRContext *context, int baseIndex) {
+  assert(barrierOp && "barrierOp requires valid operation pointer");
+  assert(context && "context requires valid MLIR context pointer");
+  uint const baseArgNum = newCircuitOp.getNumArguments();
+  int insertedCount = 0;
+  for (uint cnt = 0; cnt < barrierOp->getNumOperands(); cnt++) {
+    auto qubit = barrierOp->getOperand(cnt);
+    int argumentIndex = 0;
+    if (find(insertedArguments, cnt + baseIndex) != insertedArguments.end()) {
+      auto physicalId = qubit.getDefiningOp()->getAttrOfType<IntegerAttr>("id");
+      argumentIndex = baseArgNum + insertedCount;
+      newCircuitOp.insertArgument(baseArgNum + insertedCount, qubit.getType(),
+                                  {}, qubit.getLoc());
+      newCircuitOp.setArgAttrs(
+          argumentIndex,
+          ArrayRef({NamedAttribute(
+              StringAttr::get(context, mlir::quir::getPhysicalIdAttrName()),
+              physicalId)}));
+      insertedCount++;
+    } else {
+      argumentIndex = reusedArguments[cnt + baseIndex];
+    }
+    mapper.map(qubit, newCircuitOp.getArgument(argumentIndex));
+    baseIndex++;
+  }
+  return baseIndex;
+}
+
+void MergeCircuitsPass::mergePhysicalIdAttrs(CircuitOp newCircuitOp,
+                                             CircuitOp nextCircuitOp,
+                                             PatternRewriter &rewriter) {
+  // merge the physical ID attributes
+  auto theseIdsAttr = newCircuitOp->getAttrOfType<ArrayAttr>(
+      mlir::quir::getPhysicalIdsAttrName());
+
+  auto newIdsAttr = nextCircuitOp->getAttrOfType<ArrayAttr>(
+      mlir::quir::getPhysicalIdsAttrName());
+
+  // collect all of the first circuit's ids
+  std::vector<int> allIds;
+  for (Attribute const valAttr : theseIdsAttr) {
+    auto intAttr = valAttr.dyn_cast<IntegerAttr>();
+    allIds.push_back(intAttr.getInt());
+  }
+
+  // add IDs from the second circuit if not from the first
+  for (Attribute const valAttr : newIdsAttr) {
+    auto intAttr = valAttr.dyn_cast<IntegerAttr>();
+    auto result = std::find(begin(allIds), end(allIds), intAttr.getInt());
+    if (result == end(allIds))
+      allIds.push_back(intAttr.getInt());
+  }
```

Comment:
Same question about sets here. Set creation insertion is probably faster than this n^2 algorithm.